File: backend/core/src/modules/stats/stats.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Character, Document, Plot, WorldSetting } from '../../entities';

@Injectable()
export class StatsService {
  constructor(
    @InjectRepository(Document)
    private readonly documentRepo: Repository<Document>,
    @InjectRepository(Character)
    private readonly characterRepo: Repository<Character>,
    @InjectRepository(WorldSetting)
    private readonly worldRepo: Repository<WorldSetting>,
    @InjectRepository(Plot)
    private readonly plotRepo: Repository<Plot>,
  ) {}

  async projectStats(projectId: string) {
    const [documents, characters, worldSettings, plots] = await Promise.all([
      this.documentRepo.count({ where: { projectId } }),
      this.characterRepo.count({ where: { projectId } }),
      this.worldRepo.count({ where: { projectId } }),
      this.plotRepo.count({ where: { projectId } }),
    ]);

    const words = await this.documentRepo
      .createQueryBuilder('d')
      .select('COALESCE(SUM(d.word_count), 0)', 'sum')
      .where('d.project_id = :projectId', { projectId })
      .getRawOne();

    return {
      projectId,
      documents,
      characters,
      worldSettings,
      plots,
      wordCount: Number(words?.sum ?? 0),
    };
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/StatsService.java
package store.pjcloud.cowrite.core.service;

import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import store.pjcloud.cowrite.core.repository.CharacterRepository;
import store.pjcloud.cowrite.core.repository.DocumentRepository;
import store.pjcloud.cowrite.core.repository.PlotRepository;
import store.pjcloud.cowrite.core.repository.WorldSettingRepository;

@Service
public class StatsService {
    private final DocumentRepository documentRepo;
    private final CharacterRepository characterRepo;
    private final WorldSettingRepository worldRepo;
    private final PlotRepository plotRepo;

    public StatsService(DocumentRepository documentRepo,
                        CharacterRepository characterRepo,
                        WorldSettingRepository worldRepo,
                        PlotRepository plotRepo) {
        this.documentRepo = documentRepo;
        this.characterRepo = characterRepo;
        this.worldRepo = worldRepo;
        this.plotRepo = plotRepo;
    }

    public Map<String, Object> projectStats(UUID projectId) {
        long documents = documentRepo.findByProjectId(projectId).size();
        long characters = characterRepo.count((root, query, cb) -> cb.equal(root.get("projectId"), projectId));
        long worldSettings = worldRepo.count((root, query, cb) -> cb.equal(root.get("projectId"), projectId));
        long plots = plotRepo.count((root, query, cb) -> cb.equal(root.get("projectId"), projectId));
        Long sum = documentRepo.sumWordCount(projectId);

        return Map.of(
            "projectId", projectId,
            "documents", documents,
            "characters", characters,
            "worldSettings", worldSettings,
            "plots", plots,
            "wordCount", sum == null ? 0 : sum
        );
    }
}
